SpiffWorkflow/bpmn/specs/InclusiveGateway.py

```diff
@@ -62,58 +63,56 @@ class InclusiveGateway(UnstructuredJoin):
-        inputs_with_tokens, waiting_tasks = self._get_inputs_with_tokens(
-            my_task)
-        inputs_without_tokens = [
-            i for i in self.inputs if i not in inputs_with_tokens]
-
-        waiting_tasks = []
-        for task in tasks:
-            if (self._has_directed_path_to(
-                    task, self,
-                    without_using_sequence_flow_from=inputs_with_tokens) and
-                not self._has_directed_path_to(
-                    task, self,
-                    without_using_sequence_flow_from=inputs_without_tokens)):
-                waiting_tasks.append(task)
-
-        return force or len(waiting_tasks) == 0, waiting_tasks
-
-    def _has_directed_path_to(self, task, task_spec,
-                              without_using_sequence_flow_from=None):
-        q = deque()
-        done = set()
-
-        without_using_sequence_flow_from = set(
-            without_using_sequence_flow_from or [])
-
-        q.append(task.task_spec)
-        while q:
-            n = q.popleft()
-            if n == task_spec:
-                return True
-            for child in n.outputs:
-                if child not in done and not (
-                        n in without_using_sequence_flow_from and
-                        child == task_spec):
-                    done.add(child)
-                    q.append(child)
-        return False
+        completed_inputs, waiting_tasks = self._get_inputs_with_tokens(my_task)
+        uncompleted_inputs = [i for i in self.inputs if i not in completed_inputs]
+
+        # We only have to complete a task once for it to count, even if's on multiple paths
+        for task in waiting_tasks:
+            if task.task_spec in completed_inputs:
+                waiting_tasks.remove(task)
+
+        # If we have waiting tasks, we're obviously not done
+        # However we have to handle the case where there are paths from active tasks that must go through uncompleted inputs
+        if len(waiting_tasks) == 0:
+
+            tasks = my_task.workflow.get_tasks(TaskState.READY | TaskState.WAITING, workflow=my_task.workflow)
+            sources = [t.task_spec for t in tasks]
+
+            # This will go back through a task spec's ancestors and return the source, if applicable
+            def check(spec): 
+                for parent in spec.inputs:
+                    return parent if parent in sources else check(parent)
+
+            # If we can get to a completed input, from this task, we don't have to wait for it
+            for spec in completed_inputs:
+                source = check(spec)
+                if source is not None:
+                    sources.remove(source)
+
+            # Now check the rest of the uncompleted inputs and see if they can be reached from any of the remaining tasks
+            unfinished_paths = []
+            for spec in uncompleted_inputs:
+                if check(spec) is not None:
+                    unfinished_paths.append(spec)
```

Comment:
If force is True can this block be skipped? Like the if could be `if not force and len(waiting_tasks) == 0`. Also once you find an unfinished_path can be `complete` and break out of the loop?